sympy/simplify/_cse_diff.py

```diff
@@ -0,0 +1,263 @@
+    Jacobian matrix in DAG form.
+
+    The function also returns a list of precomputed free symbols for each subexpression,
+    which are useful in the substitution process.
+
+    NOTE: The output Jacobian might not be always exactly in a DAG form, but close to it.
+
+    Parameters
+    ==========
+
+    replacements : list of (Symbol, expression) pairs
+        Replacement Symbols and relative Common Subexpressions that have been replaced during a CSE operation.
+
+    reduced_expr : list of SymPy expressions
+        The reduced expressions with all the replacements from the replacements list above.
+
+    wrt : Matrix, tuple, or list
+        Iterable of expressions with respect to which to compute the Jacobian Matrix
+
+    Returns
+    =======
+
+    replacements : list of (Symbol, expression) pairs
+        Replacement Symbols and relative Common Subexpressions that have been replaced during a CSE operation.
+        Compared to the input replacement list, the output one doesn't contain replacement symbols inside
+        Derivative's arguments.
+
+    jacobian : list of SymPy expressions
+        The list only contains one element, which is the Jacobian Matrix with elements
+        in reduced form (replacement symbols are present)
+
+    precomputed_fs: list
+        List of sets, which store the free symbols present in each sub-expression. Useful in the substitution process.
+    """
+
+    if not isinstance(reduced_expr[0], MatrixBase):
+        raise TypeError("``expr`` must be of matrix type")
+
+    if not (reduced_expr[0].shape[0] == 1 or reduced_expr[0].shape[1] == 1):
+        raise TypeError("``expr`` must be a row or a column matrix")
+
+    if not isinstance(wrt, (MatrixBase, list, tuple)):
+        raise TypeError("``wrt`` must be an iterable of variables")
+
+    elif isinstance(wrt, (list, tuple)):
+        wrt = Matrix(wrt)
+
+    if not (wrt.shape[0] == 1 or wrt.shape[1] == 1):
+        raise TypeError("``wrt`` must be a row or a column matrix")
+
+    replacements, reduced_expr = _process_cse(replacements, reduced_expr)
+
+    if replacements:
+        rep_sym, sub_expr = map(Matrix, zip(*replacements))
+    else:
+        rep_sym, sub_expr = Matrix([]), Matrix([])
+
+    l_sub, l_wrt, l_red = len(sub_expr), len(wrt), len(reduced_expr[0])
+
+    f1 = Matrix.from_dok(l_red, l_wrt, {(i, j): diff_value for i, r in enumerate(reduced_expr[0])
+                                        for j, w in enumerate(wrt) if (diff_value := r.diff(w)) != 0})
+    if not replacements:
+        return [], [f1], []
+
+    f2 = Matrix.from_dok(l_red, l_sub, {(i, j): diff_value
+                                        for i, (r, fs) in enumerate([(r, r.free_symbols) for r in reduced_expr[0]])
+                                        for j, s in enumerate(rep_sym) if s in fs and (diff_value := r.diff(s)) != 0})
+
+    precomputed_fs = [
+        {symbol for symbol in s.free_symbols if re.compile(r'x\d+').fullmatch(symbol.name)}
```

Comment:
Why use a regular expression match? This may result in errors when someone uses a different generator in cse. I would expect that you can create a set with all replacement symbols and take the intersection here.